plutus-ir/src/Language/PlutusIR/Compiler/Let.hs

```diff
@@ -62,45 +62,47 @@ compileLet :: Compiling m e uni a => LetKind -> PIRTerm uni a -> DefT SharedName
 compileLet kind = \case
     Let p r bs body -> withEnclosing (const $ LetBinding r p) $ case r of
             -- See Note [Right-associative compilation of let-bindings for linear scoping]
-            NonRec -> lift $ foldM (compileNonRecBinding kind) body (reverse bs)
+            NonRec -> lift $ foldM (compileNonRecBinding kind) body (NE.reverse bs)
             Rec    -> compileRecBindings kind body bs
     x -> pure x
 
 compileRecBindings
     :: Compiling m e uni a
     => LetKind
     -> PIRTerm uni a
-    -> [Binding TyName Name uni (Provenance a)]
+    -> NE.NonEmpty (Binding TyName Name uni (Provenance a))
     -> DefT SharedName uni (Provenance a) m (PIRTerm uni a)
-compileRecBindings kind body bs
-    | null typeBinds = compileRecTermBindings kind body termBinds
-    | null termBinds = lift $ compileRecTypeBindings kind body typeBinds
-    | otherwise      = lift $ getEnclosing >>= \p -> throwing _Error $ CompilationError p "Mixed term and type bindings in recursive let"
+compileRecBindings kind body bs =
+  if null mTailGroups
```

Comment:
Style tip: you almost never want to use `null`, pattern match instead. In this case, just name the output of the function on line 84, and match on it here:
```
case grouped of
    group :| [] -> ...
    _ :| xs -> ...
```

Reply:
Generalizing a little: if you find yourself writing `if <condition>` where the condition just checks which branch of the datatype you're in, you almost always want to write it as a pattern match instead.